CORE-308: Track Tokens Dynamically

chrisnevers opened the pull request.

jeapostrophe commented on examples/abstract-tok/index.rsh:
```diff
@@ -0,0 +1,44 @@
+'reach 0.1';
+
+const MT = Maybe(Token);
```
I think another good test would be making it so the HumbleSwap code is actually the same but with a `Maybe(Token)` parameter for the A side

chrisnevers replied:
I don't think it's possible still because `paySpec` requires a hardcoded array of identifiers

jeapostrophe replied:
Could you write `.paySpec( mTokA.match({None: xyz, Some: abc}) )`? I thought that was an evaluated position?

chrisnevers replied:
No, it's not evaluated. We use the `JSExpression` directly to generate code in `doFork`

jeapostrophe replied:
Well, maybe we should evaluate it and then synthesize something new in its place :) If not, then we should at least write down the examples that fail/are-not-allowed because it is not evaluated

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -5465,6 +5479,9 @@ evalStmt = \case
             let vv = case vt of
                   T_Null -> SLV_Null at_c "case"
+                  T_Token -> case de_val of
+                      SLV_Data _ _ _ (SLV_DLTok (DLToken _ i)) -> SLV_DLTok (DLToken dv' i)
```
This line shows the brittleness of this technique. I think if you write more test scenarios, you'll see problems with it. For example... a tuple of two tokens, an array of tokens, a function that returns a token based on some condition, a token sent in to the program that is identical to an earlier token (asserted with a require)

I think you could generalize and improve your technique a little bit by making the `ST_Token` type have a `(Maybe Int)` parameter which is used to identify which it is and if it is `Nothing`, then that means "any token". I'm not sure how our type machinery will work with that, but I hypothesize it will be easier. I don't think that the DL types would change at all, but maybe I'm wrong because of the way balance assertions work.

I think the most powerful technique will be "valued based". Remember, the reason we need to keep track of these things is so that we can (a) track the balance adjustments and (b) track the balances in loops with the fluid variable capture. The dumbest thing that could possibly work would be (1) representing the balances as `Map`-like type in DL, (2) using SMT's maps for that during verification, and (3) compiling it to an array on a platform like ALGO that doesn't have maps (and maybe it would be good to be an array on others when we know the identity of tokens statically.) This would basically be like adding a new `T_TokenBalances` type.

Maybe my two ideas are too heavy, but I don't think we can evaluate your idea without more examples of what it allows and doesn't allow.

chrisnevers replied:
I get why this seems brittle, but in this case, we know that if the unboxed value being bound by a pattern match is of type `T_Token`, it can only be `SLV_Data _ _ _ (SLV_DLTok (DLToken _ i))`.

I will update the example to demonstrate how it works with arrays, functions etc. so long as it's statically computable. When a token sent in to the program is identical to an earlier token, it errors as expected.

I think the DL `T_Token` type would require the `Maybe (Int)` information for looking up the balances. In my opinion, this approach could get ugly later in the compiler when we have Map keys that are `T_Token`, so we'd probably have to intercept lookups to treat `lookup (T_Token (Just ..)) map` as `lookup (T_Token Nothing) map`

I kind of get the gist of the value-based approach, but I don't think I understand it well. For

> (1) representing the balances as Map-like type in DL

Would Tokens be `DLVar`s with type: `T_TokenBalances (Map DLVar Int)`?

jeapostrophe replied:
Let's break this conversation up into two pieces.

First, I want more systematic and small tests (okay if they are just in `hs/t` and compile-time only) that show what is possible _now_, with *this* implementation, and what is still not possible.

The way I'd describe in words is that today, tokens must appear at the top-level of a publish and must remain completely un-abstracted in every operation, thus they must be compile-time static values the entire time. So, you can't publish a Maybe Token and expect it to be visible, you have to publish the actual token. Inside one consensus step, you can't stick two tokens into a container and then pull them out and then be able to use them afterwards. What about across a consensus step... does that matter?

With this change, I think that token identities are preserved in DLArgs, but I don't have a good sense of what that allows or doesn't.

--

Second, the different implementations.

About the type idea... you have introduced an argument `DLToken DLVar Int` with the invariant that `DLVar` is always `DLVar src lab T_Token id`. My point is just that `DLToken (DLVar src lab T_Token id) which` is equivalent to storing the `which` in the token type `DLVar src lab (T_Token (Just which)) id` and is like the common dependent type concept of an indexed type, so I think it would fit in better if you did it that way. (You need the `Maybe`, because in your version some `DLVar`s that are `T_Token` aren't wrapped in a `DLToken`.)

About the map idea... what's the point of keeping track of tokens? It is so that when you do `transfer(x, tok)` we can do `(assert (> (balance tok) x))` and `(fluid-set (balance tok) (- (balance tok) x)`. Everything about tokens is all about that little snippet. Right now, we do that by implementing `(balance tok)` with a fixed array that is referenced by position which requires us to know which index `tok` is at. It's like its type is `Array(UInt, HowManyTokensInProgram)`. We could instead represent it with `Array(Tuple(Token, UInt), HowManyTokensInProgram)` and then implement (balance tok)` with a simple search. In SMT, we could compile that "search" with an actual map. With some clever compilation, we may also be able to get the exact position a lot of the time. I have an idea for how to implement this smoothly that I can explain also, but I think I'd rather speak it than write it.

jeapostrophe commented on examples/abstract-tok/index.mjs:
```diff
@@ -11,12 +11,14 @@ const startingBalance = pc(100);
 
   const zmd = await stdlib.launchToken(accAlice, "Zorkmid", "ZMD");
   const gil = await stdlib.launchToken(accAlice, "Gil", "GIL");
+  const gol = await stdlib.launchToken(accAlice, "Gold", "AU");
```
I think gold would be GP ;)

jeapostrophe commented on examples/abstract-tok/index.rsh:
```diff
@@ -29,16 +30,32 @@ export const main = Reach.App(() => {
 
-  go(MS(token1));
+  const ta = array(Token, [ getTok(4), token2 ]);
+  go(MS(ta));
```
This is better

jeapostrophe commented on hs/t/n/Err_InspectAbstractToken.rsh:
```diff
@@ -0,0 +1,18 @@
+  A.publish(mt);
+  commit();
+
+  A.pay([ maybe(mt, 0, (tok) => [ 5, tok ]) ]);
```
What does this program do without your other changes? My assumption is that your code makes it so I can't even write
```
const x = fromMaybe(maybeTok, otherTok);
view.set(x);
```
because I would "inspecting" a token from a container. But the actual problem is that we aren't allowed to be paid in or pay with a token that we learned this way.

jeapostrophe commented on hs/t/y/abs_tok_container.rsh:
```diff
@@ -0,0 +1,26 @@
+'reach 0.1';
```
This fails w/o your patch?

jeapostrophe commented on hs/t/y/abs_tok_fun.rsh:
```diff
@@ -0,0 +1,33 @@
+'reach 0.1';
```
Ditto?

chrisnevers commented on hs/src/Reach/Verify/SMT.hs:
```diff
@@ -711,13 +712,20 @@ pathAddUnbound at_dv (Just dv) msmte = do
   v <- smtVar dv
   let smlet = Just . SMTLet at_dv dv (DLV_Let DVC_Once dv) sc =<< de
   smtDeclare_v v t smlet
+  assertInvariants at_dv t v
```
@jeapostrophe  Any harm in this? Otherwise, making claims about balances in loop invariants require `balance(token) <= UInt.max`

jeapostrophe replied:
No harm

chrisnevers commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -639,6 +638,7 @@ data DLExpr
   | DLE_GetUntrackedFunds SrcLoc (Maybe DLArg) DLArg
   | DLE_FromSome SrcLoc DLArg DLArg
   -- Maybe try to generalize FromSome into a Match
+  | DLE_BalanceInit Int DLVar
```
This is for the initial creation of the balance array. At the start of the program, we don't know how big the array will be yet, so this will transform to an actual array after evaluation. DLVar is `<None null>` which will be used to initialize every spot in the array.

jeapostrophe replied:
Why are they maybes, rather than just UInts? And what's the number?

chrisnevers commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -55,12 +55,16 @@ data DLType
   | T_Object (M.Map SLVar DLType)
   | T_Data (M.Map SLVar DLType)
   | T_Struct [(SLVar, DLType)]
+  | T_TokenBalances Integer
```
Don't know the size of the array until after evaluation; this will transform to the actual array type

jeapostrophe replied:
What's the integer?

chrisnevers commented on hs/src/Reach/Linearize.hs:
```diff
@@ -338,15 +338,23 @@ data DFEnv = DFEnv
+adjustFluids = \case
+  FV_balances _ _ -> do
+    eBals <- asks eBals
+    return $ FV_balances eBals mempty
```
I temporarily did all the work of tweaking the balance values/types to the correct array size in Linearize just to see it work. I think I want to make a new pass immediately after evaluation

jeapostrophe commented on hs/rsh/stdlib.rsh:
```diff
@@ -166,6 +166,9 @@ export const Array_indexOf =
   (arr) => (x) => Array_indexOf(arr, x);
 
+export const Array_set =
+  (x, y, z) => Array.set(x, y, z)
```
Why?

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -224,6 +229,23 @@ litTypeOf = \case
   DLL_Bool _ -> T_Bool
   DLL_Int {} -> T_UInt
 
+data DLToken = DLToken DLVar Int
```
Don't you not need this anymore?

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -1058,7 +1088,7 @@ instance Pretty a => Pretty (DLRecv a) where
 
 data FluidVar
-  = FV_balance Int
+  = FV_balances Integer (M.Map DLArg Int)
```
What's the integer and the map? I thought this was going to just be `FV_balances` (no parameter)

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -1083,9 +1113,12 @@ instance Pretty FluidVar where
 fluidVarType :: FluidVar -> DLType
 fluidVarType = \case
-  FV_balance _ -> T_UInt
+  FV_balances i _ -> T_TokenBalances (i + 1)
```
I thought that `FV_balances` would be of type `T_TokenBalances` (no parameter) or we could manage to avoid looking at it until later and thus never need to see the type, so this would be `impossible`

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2184,13 +2184,117 @@ doBalanceInit_ fv mtok amta = do
+  _ -> impossible "getBalsEnv: Expected FV_balances"
+
+networkBal :: SrcLoc -> SLVal
+networkBal at = SLV_Data at maybeTokenMap "None" $ SLV_Null at "networkBalance"
```
Ah, I understand why you have the maybe now. I think it will be simpler to just have two paths... one for the network and one for the non-network. All these maybes are complex

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2184,13 +2184,117 @@ doBalanceInit_ fv mtok amta = do
+            return r
+    Just ow -> impossible $ "getBalanceIndex: tsv received: " <> show ow
+
+setBalance :: Maybe DLArg -> DLArg -> ReaderT Env IO DLArg
```
Use `App`

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2184,13 +2184,117 @@ doBalanceInit_ fv mtok amta = do
+        case join $ fmap (flip M.lookup env) mtok of
+          Just i -> return $ public $ SLV_Int at $ fromIntegral i
+          Nothing -> do
+            b <- doFluidRef bals
```
Ah, so this expands to an array of at least as big as the real thing at this point in the program

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2184,13 +2184,117 @@ doBalanceInit_ fv mtok amta = do
+    Just (DLA_Var dv) -> do
+        let env = getBalsEnv bals
+        case join $ fmap (flip M.lookup env) mtok of
+          Just i -> return $ public $ SLV_Int at $ fromIntegral i
```
And this path is looking for when it is fixed and happens to be static

jeapostrophe commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2205,7 +2309,22 @@ doBalanceAssert_ fv mtok lhs op msg = do
+    -- Set value/fluid ref with updated array
+    b' <- setBalance mtok bva
+    doFluidSet_ b b'
+
```
This whole approach is interesting and different than I was thinking. I was thinking there would be `DLE_BalanceGet` and `DLE_BalanceSet` and `T_Balances` and then `FV_balances` of type `T_Balances` and then we'd transform them. But, thinking now, we might even get away with making `DLS_BalanceInit SrcLoc`, `DLS_BalanceGet SrcLoc DLVar DLArg`, and `DLS_BalanceSet SrcLoc DLVar DLArg DLArg` and then expand those into the array of the right size and uses of fluid set and ref. Then the rest of the compiler doesn't need to think about them at all and it happens before Linearize and we can discover the correct size right away.

chrisnevers replied:
I will try this approach/what you described in slack

jeapostrophe commented on hs/src/Reach/Eval/Error.hs:
```diff
@@ -727,5 +729,7 @@ instance Show EvalError where
     Err_ForkNoCases ->
       "A `fork`/`parallelReduce` statement must include a `.case` or `.api` component"
+    Err_InspectAbstractToken ->
+      "Cannot inspect `Token` that was published inside of a container."
```
Do you still throw this error? I don't think so?

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -193,19 +197,26 @@ instance Pretty DLInit where
 data DLConstant
   = DLC_UInt_max
-  deriving (Eq, Generic, Show, Ord)
+  | DLC_Zero_addr
```
It's not really a zero address... it is a zero token, which would be `DLC_Token_zero`

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -55,15 +55,19 @@ data DLType
   | T_Data (M.Map SLVar DLType)
   | T_Struct [(SLVar, DLType)]
-  | T_TokenBalances Integer
+  | T_Balances
```
Because of the way that `fluidVarType` is used, I don't think you actually need to define this, provided you special case it in the `While` part of `Linearize` with the actual type. This will simplify things greatly.

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -112,7 +112,9 @@ conName' :: T.Text
-conCons' DLC_UInt_max = DLL_Int sb $ 2 ^ (64 :: Integer) - 1
+conCons' = \case
+  DLC_UInt_max  -> DLL_Int sb $ 2 ^ (64 :: Integer) - 1
+  DLC_Zero_addr -> DLL_Int sb $ 0
```
This is not really correct... Right now the code makes the assumption that "constants can be turned into literals", but this conversion would be of the wrong type, because `Token` is a different type than `Int`

jeapostrophe commented on hs/src/Reach/Connector/ETH_Solidity.hs:
```diff
@@ -57,7 +57,9 @@ conName' :: T.Text
-conCons' DLC_UInt_max = DLL_Int sb $ 2 ^ (256 :: Integer) - 1
+conCons' = \case
+  DLC_UInt_max  -> DLL_Int sb $ 2 ^ (256 :: Integer) - 1
+  DLC_Zero_addr -> impossible "ETH conCons': zero_addr"
```
See the comment about, but in terms of code generation it is possible... `address(0)`

jeapostrophe commented on hs/src/Reach/Connector/ETH_Solidity.hs:
```diff
@@ -478,6 +480,7 @@ solLit = \case
 solArg :: AppT DLArg
 solArg = \case
   DLA_Var v -> solVar v
+  DLA_Constant DLC_Zero_addr -> return "payable(address(0))"
```
Ah, you know that

jeapostrophe commented on hs/src/Reach/Linearize.hs:
```diff
@@ -532,6 +573,19 @@ df_eb :: DKExportBlock -> DFApp DLExportBlock
+  let false = DLA_Literal $ DLL_Bool False
+  let zero = DLA_Literal $ DLL_Int sb 0
+  let c1 = DKC_Let sb (DLV_Let DVC_Many nv) $ DLE_LArg sb $ DLLA_Tuple [DLA_Constant DLC_Zero_addr, zero, zero, false]
+  let c2 = DKC_Let sb lv $ DLE_BalanceInit nv
```
I don't understand what `DLE_BalanceInit` is for

chrisnevers replied:
I will remove it as I can just do the conversion to an array directly in this function

jeapostrophe commented on hs/src/Reach/Simulator/Core.hs:
```diff
@@ -363,7 +363,9 @@ interpPrim = \case
-conCons' DLC_UInt_max = V_UInt $ 2 ^ (64 :: Integer) - 1
+conCons' = \case
+  DLC_UInt_max  -> V_UInt $ 2 ^ (64 :: Integer) - 1
+  DLC_Zero_addr -> V_Address 0
```
It's a token

chrisnevers commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -4291,7 +4326,28 @@ compilePayAmt tt v = do
+                    tup <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing tupTy) $ DLE_TupleRef at (DLA_Var dv) idx)
+                    amt_a <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing T_UInt) $ DLE_TupleRef at tup 0)
+                    token_a <- DLA_Var <$> (ctxt_lift_expr (DLVar at Nothing T_UInt) $ DLE_TupleRef at tup 1)
+                    when (token_a `elem` sks) $ do
```
@jeapostrophe Would verifying that all the tokens in the PayAmt are unique be left up to the programmer? For example, if there are two dynamic token references, would something like the following be required:

```Javascript
const tokA = // dynamic computation
const tokB = // dynamic computation
A.pay([ [1, tokA], [1, tokB] ]).payRequire(() => {
  require(tokA != tokB);
);
```

jeapostrophe replied:
First, for now, you'd generate an `assert` and then the user is responsible for adding the appropriate `assume`/`require`. Meaning `Err_Transfer_DoubleToken` goes away and is replaced by a verification error

Second, ideally we'd be able to remove this constraint all together and support examples where the same token is given multiple times.

jeapostrophe commented on hs/src/Reach/Linearize.hs:
```diff
@@ -389,88 +389,114 @@ expandFromFVMap at (DLAssignment updatem) = do
+  let bl = DLBlock at [] ss $ DLA_Var bl_res
+  let s0 = asn init_acc_dv $ DLE_LArg at $ DLLA_Tuple [DLA_Literal $ DLL_Bool False, DLA_Literal $ DLL_Int at 0]
+  let s1 = DL_ArrayReduce at reduce_res bals (DLA_Var init_acc_dv) acc_dv elem_dv bl
+  -- XXX add check to die if `is_found` false?
```
Add an `assert` that it is `true`

jeapostrophe commented on hs/src/Reach/Linearize.hs:
```diff
@@ -389,88 +389,114 @@ expandFromFVMap at (DLAssignment updatem) = do
     supply <- mkVar at $ T_UInt
     destroyed <- mkVar at $ T_Bool
     let c1 = asn tup $ DLE_ArrayRef at bals idx
     -- XXX Make DLE_TupleSet? This is kind of gross
```
Probably

jeapostrophe commented on hs/src/Reach/Optimize.hs:
```diff
@@ -599,11 +599,12 @@ instance Optimize DLStmt where
-      DL_ArrayReduce at ans <$> opt x <*> opt z <*> (pure b) <*> (pure a) <*> opt f
+    DL_ArrayMap at ans x a i f -> do
+      s' <- DL_ArrayMap at ans <$> opt x <*> pure a <*> pure i <*> opt f
+      maybeUnroll s' x $ return s'
```
Whoops! This is great

jeapostrophe commented on hs/t/n/nonunique_payamt.rsh:
```diff
@@ -0,0 +1,31 @@
+'reach 0.1';
```
Missing `.txt`

chrisnevers commented:
@jeapostrophe  Some verification issues arose when I added the assertion that a dynamic token lookup must be successful. 
The issue is that the `tokenInfo` array can be a loop variable, so its token keys "could" change. A reach program can't express that the token keys will be static and I felt like enforcing the SMT to treat the `tokenInfo` array specially would be a large undertaking. 

So instead, I split the token references into their own array:

```Javascript
tokenInfo: Array([UInt, UInt, Bool], <size>);
tokens : Array(Token, <size>);
```

I explicitly prevent `tokens` from becoming a loop variable. Now, the token array is static once initialized. However, since it can't become a loop variable we still have the restriction that you can't publish tokens within a loop.

chrisnevers commented on hs/src/Reach/Eval/Core.hs:
```diff
@@ -2387,15 +2426,20 @@ evalPrim p sargs =
       tokdv <- doInternalLog_ Nothing tokdv_
       st <- readSt id
+      let existingToks = st_toks st
+      tokIsUniq <- tokIsUnique (map DLA_Var existingToks) $ DLA_Var tokdv
```
When creating a token, assume that it's unique from all the existing tokens. This is needed for verifying that the tokens in a pay amount are unique

jeapostrophe replied:
This is a valid assumption too :)

jeapostrophe commented on hs/t/n/Err_Transfer_DoubleToken.rsh:
```diff
@@ -1,20 +0,0 @@
-'reach 0.1';
```
We should never delete tests, but we can move them or rename them

jeapostrophe commented on hs/t/n/mint-fail1.txt:
```diff
@@ -9,23 +9,133 @@ Verification failed:
   // Theorem Formalization
 
-  const v362 = 0 == v257.supply;
+  const v307 = [0, 0, false ];
```
This is really confusing... is it plausible to insert in some labels on the variables in your variable making function?

jeapostrophe reviewed (changes requested):
Can you also check that the cost of Humble doesn't change?

chrisnevers commented:
Ready to go, duoswap is now 707 cost though

jeapostrophe reviewed (commented):
.
